File: demos/hotword/lattigo/evaluate_fhe.go
// Evaluate a single inference
package main

import (
	"flag"
	"fmt"
	"os"
	"time"

	"fully_homomorphic_encryption/demos/common/go/pathutils"
	"fully_homomorphic_encryption/demos/hotword/lattigo/hotword_lattigo"
	"fully_homomorphic_encryption/demos/hotword/lattigo/hotword_lattigo_utils"
)

var labels = [12]string{
	"_silence_",
	"_unknown_",
	"yes",
	"no",
	"up",
	"down",
	"left",
	"right",
	"on",
	"off",
	"stop",
	"go",
}

func main() {
	sampleIdxFlag := flag.Int("sample_idx", 0, "Sample index in the NPZ to test")
	npzPathFlag := flag.String("npz_path", "test_data.npz", "Path to the test NPZ file")
	flag.Parse()

	npzPath := *npzPathFlag
	if npzPath == "test_data.npz" {
		npzPath = pathutils.ResolvePath("fully_homomorphic_encryption/demos/hotword/data/test_data.npz")
	}
	sampleIdx := *sampleIdxFlag

	fmt.Printf("Loading test sample %d from %s...\n", sampleIdx, npzPath)
	t0 := time.Now()
	features, expectedLabel, err := loadTestRow(npzPath, sampleIdx)
	if err != nil {
		fmt.Printf("Error loading test row: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("  Took %v\n", time.Since(t0))
	fmt.Printf("  Expected label: %s (%d)\n", labels[expectedLabel], expectedLabel)
	fmt.Printf("  Feature vector size: %d\n", len(features))

	// Configure context
	fmt.Println("Configuring Lattigo context...")
	t0 = time.Now()
	btpEvaluator, evaluator, params, ecd, encryptor, decryptor := hotword_lattigo.Tcresnet8small__configure()
	fmt.Printf("  Took %v\n", time.Since(t0))

	// Encrypt input
	fmt.Println("Encrypting input features...")
	t0 = time.Now()
	encryptedFeatures := hotword_lattigo.Tcresnet8small__encrypt__arg0(evaluator, params, ecd, encryptor, features)
	fmt.Printf("  Took %v\n", time.Since(t0))

	// Preprocessing
	fmt.Println("Running preprocessing...")
	t0 = time.Now()
	preprocessedWeights := hotword_lattigo_utils.Tcresnet8small__preprocessing(params, ecd)
	fmt.Printf("  Took %v\n", time.Since(t0))

	// FHE evaluation
	fmt.Println("Running FHE evaluation (preprocessed)...")
	t0 = time.Now()
	ctZeros := hotword_lattigo.Tcresnet8small__encrypt__zeros(evaluator, params, ecd, encryptor)

	encryptedOutput := hotword_lattigo.Tcresnet8small__preprocessed(
		btpEvaluator, evaluator, params, ecd, encryptedFeatures,
		ctZeros,
		preprocessedWeights,
	)
	fmt.Printf("  Took %v\n", time.Since(t0))

	// Decrypt
	fmt.Println("Decrypting output...")
	t0 = time.Now()
	decryptedLogits := hotword_lattigo.Tcresnet8small__decrypt__result0(evaluator, params, ecd, decryptor, encryptedOutput)
	fmt.Printf("  Took %v\n", time.Since(t0))

	fmt.Printf("Decrypted logits: %v\n", decryptedLogits)

	// Argmax
	predictedClass := 0
	maxVal := decryptedLogits[0]
	for i := 1; i < 12; i++ {
		if decryptedLogits[i] > maxVal {
			maxVal = decryptedLogits[i]
			predictedClass = i
		}
	}
	fmt.Printf("Predicted class: %s (%d)\n", labels[predictedClass], predictedClass)

	if predictedClass == expectedLabel {
		fmt.Println("SUCCESS: Predicted class matches expected label!")
	} else {
		fmt.Println("FAILURE: Predicted class does NOT match expected label!")
		os.Exit(1)
	}
}

File: demos/cc_fraud/openfhe/evaluate_fhe.py
"""Evaluate the fraud detection model using OpenFHE."""

import argparse
import os
import time

import numpy as np
import pandas as pd

from demos.cc_fraud.openfhe import fraud_model_pybind
from demos.cc_fraud.utils.data_utils import load_test_row
from demos.common.python import path_utils

resolve_path = path_utils.resolve_path


def main():
  parser = argparse.ArgumentParser()
  parser.add_argument(
      "--row_idx",
      type=int,
      default=0,
      help="Row index from test_rows.csv to evaluate",
  )
  parser.add_argument("--csv_path", type=str, default="test_rows.csv")
  args = parser.parse_args()

  csv_path = args.csv_path
  if csv_path == "test_rows.csv":
    csv_path = resolve_path(
        "demos/cc_fraud/data/test_rows.csv"
    )

  print(f"Loading test row {args.row_idx} from {csv_path}...")
  t0 = time.time()
  features, expected_label = load_test_row(csv_path, args.row_idx)
  print(f"  Took {time.time() - t0:.4f} seconds")
  print(f"  Expected label (is_fraud): {expected_label}")
  print(f"  Feature vector size: {len(features)}")
  print(f"  First 5 features: {features[:5]}")

  # Initialize crypto context
  print("Generating crypto context...")
  t0 = time.time()
  cc = fraud_model_pybind.cc_fraud__generate_crypto_context()
  print(f"  Took {time.time() - t0:.4f} seconds")

  print("Generating key pair...")
  t0 = time.time()
  key_pair = cc.KeyGen()
  public_key = key_pair.publicKey
  secret_key = key_pair.secretKey
  print(f"  Took {time.time() - t0:.4f} seconds")

  print("Configuring crypto context...")
  t0 = time.time()
  cc = fraud_model_pybind.cc_fraud__configure_crypto_context(cc, secret_key)
  print(f"  Took {time.time() - t0:.4f} seconds")

  # Encrypt input features
  print("Encrypting input features...")
  t0 = time.time()
  encrypted_features = fraud_model_pybind.cc_fraud__encrypt__arg0(
      cc, features, public_key
  )
  print(f"  Took {time.time() - t0:.4f} seconds")

  # Preprocessing
  print("Running preprocessing...")
  t0 = time.time()
  prep_struct = fraud_model_pybind.cc_fraud__preprocessing(cc)
  print(f"  Took {time.time() - t0:.4f} seconds")

  # Call the FHE function
  print("Running FHE evaluation (preprocessed)...")
  t0 = time.time()
  ct_zero_1 = fraud_model_pybind.cc_fraud__encrypt__zero__0(cc, public_key)
  ct_zero_2 = fraud_model_pybind.cc_fraud__encrypt__zero__1(cc, public_key)
  encrypted_output = fraud_model_pybind.cc_fraud__preprocessed(
      cc,
      encrypted_features,
      ct_zero_1,
      ct_zero_2,
      prep_struct,
  )
  print(f"  Took {time.time() - t0:.4f} seconds")

  # Decrypt output
  print("Decrypting output...")
  t0 = time.time()
  decrypted_logits = fraud_model_pybind.cc_fraud__decrypt__result0(
      cc, encrypted_output, secret_key
  )
  print(f"  Took {time.time() - t0:.4f} seconds")

  # The output dimension of fraud_model is tensor<1x2xf32>, representing
  # two logits.
  print(f"Decrypted logits: {decrypted_logits}")

  predicted_class = int(np.argmax(decrypted_logits))
  print(f"Predicted class: {predicted_class}")

  if predicted_class == expected_label:
    print("SUCCESS: Predicted class matches expected label!")
  else:
    print("FAILURE: Predicted class does NOT match expected label!")


if __name__ == "__main__":
  main()
